crates/amalthea/src/kernel.rs

```diff
@@ -78,418 +55,563 @@ pub enum StreamBehavior {
     None,
 }
 
-impl Kernel {
-    /// Create a new Kernel, given a connection file from a frontend.
-    pub fn new(name: &str, file: ConnectionFile) -> Result<Kernel, Error> {
-        let key = file.key.clone();
-
-        let (iopub_tx, iopub_rx) = bounded::<IOPubMessage>(10);
-
-        // Create the pair of channels that will be used to relay messages from
-        // the open comms
-        let (comm_manager_tx, comm_manager_rx) = bounded::<CommManagerEvent>(10);
-
-        Ok(Self {
-            name: name.to_string(),
-            connection: file,
-            session: Session::create(key)?,
-            iopub_tx,
-            iopub_rx: Some(iopub_rx),
+/// Connects the Kernel to the frontend
+pub fn connect(
+    name: &str,
+    connection_file: ConnectionFile,
+    registration_file: Option<RegistrationFile>,
+    shell_handler: Arc<Mutex<dyn ShellHandler>>,
+    control_handler: Arc<Mutex<dyn ControlHandler>>,
+    lsp_handler: Option<Arc<Mutex<dyn ServerHandler>>>,
+    dap_handler: Option<Arc<Mutex<dyn ServerHandler>>>,
+    stream_behavior: StreamBehavior,
+    iopub_tx: Sender<IOPubMessage>,
+    iopub_rx: Receiver<IOPubMessage>,
+    comm_manager_tx: Sender<CommManagerEvent>,
+    comm_manager_rx: Receiver<CommManagerEvent>,
+    // Receiver channel for the stdin socket; when input is needed, the
+    // language runtime can request it by sending an StdInRequest::Input to
+    // this channel. The frontend will prompt the user for input and
+    // the reply will be delivered via `stdin_reply_tx`.
+    // https://jupyter-client.readthedocs.io/en/stable/messaging.html#messages-on-the-stdin-router-dealer-channel.
+    // Note that we've extended the StdIn socket to support synchronous requests
+    // from a comm, see `StdInRequest::Comm`.
+    stdin_request_rx: Receiver<StdInRequest>,
+    // Transmission channel for StdIn replies
+    stdin_reply_tx: Sender<crate::Result<InputReply>>,
+) -> Result<(), Error> {
+    let ctx = zmq::Context::new();
+
+    let session = Session::create(connection_file.key.as_str())?;
+
+    // Channels for communication of outbound messages between the
+    // socket threads and the 0MQ forwarding thread
+    let (outbound_tx, outbound_rx) = unbounded();
+
+    // Create the comm manager thread
+    CommManager::start(iopub_tx.clone(), comm_manager_rx);
+
+    // Create the Shell ROUTER/DEALER socket and start a thread to listen
+    // for client messages.
+    let shell_socket = Socket::new(
+        session.clone(),
+        ctx.clone(),
+        String::from("Shell"),
+        zmq::ROUTER,
+        None,
+        connection_file.endpoint(connection_file.shell_port),
+    )?;
+    let shell_port = port_finalize(&shell_socket, connection_file.shell_port)?;
```

Comment:
Since the port might be chosen randomly by the OS, depending on the details of the connection file, we now potentially have to retrieve the port from the socket with `port_finalize()`.